src/box/lua/config/applier/credentials.lua

```diff
@@ -525,17 +525,93 @@ end
 local privileges_action_f = function(grant_or_revoke, role_or_user, name, privs,
                                      obj_type, obj_name)
     assert(grant_or_revoke == 'grant' or grant_or_revoke == 'revoke')
+    privs = table.copy(privs)
+    table.sort(privs)
     privs = table.concat(privs, ',')
 
     -- Try to apply the action immediately. If the object doesn't exist,
     -- the sync will be applied inside the trigger on object creation/rename.
     local ok, err = pcall(box.schema[role_or_user][grant_or_revoke],
                           name, privs, obj_type, obj_name)
 
+    -- This key identifies an alert to be dropped on a successful
+    -- operation and to be set on an error.
+    --
+    -- Let's consider the following situation: a space does not
+    -- exist, but the configuration contains privileges for it
+    -- (for example 'read' and 'write'). The warning will be
+    -- issued with the "box.schema.user.grant(<...>, 'read,write',
+    -- <...>)" key.
+    --
+    -- Now, let's assume that the space is created. We should drop
+    -- the warning and, so, we should generate exactly same key to
+    -- pass to the :drop() aboard method.
+    --
+    -- 1. Is it possible that the key to drop contains just 'read',
+    --    not 'read,write'? Seems unlikely.
```

Comment:
This comment is somewhat confusing to me. I feel that it should describe the guarantees that we rely on to provide the exact same warning message to the `aboard`. Instead it goes over some corner cases and solves them with a varying degree of confidence ("Seems, unlikely", "Seems, no").

The sync (that does the grants and revokes) is called in two scenarios:
1) on reload, when config is changed and all of such warnings are dropped anyway
2) on object appearance, when the config remains the same

The second one might be an issue, because an object can be renamed to a configurable one. At this point it might already have the permissions that are to be granted. Than, `privileges_action_f()` isn't called, because the diff is empty and no warnings are dropped.

Reply:
E.g. the config includes the following:

```yaml
privileges:
- permissions: [read, write]
  spaces: [myspace1]
```

1) there is no `myspace1`
2) `myspace2` is created and `myuser` is granted `read,write` for it
3) `myspace2` is renamed to `myspace1`

This way, the applier does not grant/revoke a thing and the configuration is still in sync.

Reply:
Looks valid. Thanks! I'll think around.

Reply:
I've implemented another solution that covers this scenario. See details in the description of the last commit.

In brief, a simple way with discarding an alert after granting of a privilege doesn't work here. We need to traverse alerts on an upper level and compare them to an actual state of privileges in the database.

I've tried to implement the solution in a more or less simple way. It means several scans over the alerts. I think that it shouldn't be a problem: the cost is anyway small comparing to a write of the rename operation to WAL. And it shouldn't be a frequent operation.

Reply:
The solution looks valid, but quite complicated. It seems like you are trying to store and update independent diff between current configuration and desired one (after a sync, we don't consider in-progress situation). As the alerts don't have to be exactly up to date at each moment, maybe it is fine to calculate this diff once in a while? We can mark all grant warning in a separate subtype, drop them all together and add new base on the diff (using  `privileges_subtract()`).

Reply:
The solution looks valid, but quite complicated. It seems like you are trying to store and update independent diff between current configuration and desired one (after a sync, we don't consider in-progress situation). As the alerts don't have to be exactly up to date at each moment, maybe it is fine to calculate this diff once in a while? We can mark all grant warning in a separate subtype, drop them all together and add new base on the diff (using  `privileges_subtract()`).

Reply:
This can also simplify the status state machine implementation. The state would change only on this diff calculation.

Reply:
I looked into it and came up with a draft diff:
```diff
diff --git a/src/box/lua/config/applier/credentials.lua b/src/box/lua/config/applier/credentials.lua
index c2fb2cba8..553049a88 100644
--- a/src/box/lua/config/applier/credentials.lua
+++ b/src/box/lua/config/applier/credentials.lua
@@ -654,6 +654,8 @@ local function sync_privileges(credentials, obj_to_sync)
                     obj_to_sync.type, obj_to_sync.name)
     end
 
+    aboard_drop_alerts({trait = 'postponed_grant'})
+
     -- The privileges synchronization between A and B is performed in 3 steps:
     -- 1. Grant all privileges that are present in B,
     --    but not present in A (`grant(B - A)`).
@@ -703,7 +705,22 @@ local function sync_privileges(credentials, obj_to_sync)
             end
         end
 
-        sync_alerts(role_or_user, name)
+        -- Update box_privileges to check which grants have been postponed and
+        -- issue a warning for each one of them.
+        box_privileges = box.schema[role_or_user].info(name)
+        box_privileges = privileges_from_box(box_privileges)
+        local postponed_grants = privileges_subtract(config_privileges, box_privileges)
+
+        for _, grant in ipairs(postponed_grants) do
+            local alert = {}
+            local msg = "credentials.apply: %s %q hasn't been created yet, " ..
+                        "'box.schema.%s.%s(%q, %q, %q, %q)' will be applied later"
+            alert.type = 'warn'
+            alert._trait = 'postponed_grant'
+            alert.message = msg:format(grant.obj_type, grant.obj_name, role_or_user,
+                                       'grant', name, grant.privs, grant.obj_type,
+                                       grant.obj_name)
+            config._aboard:set(alert)
+        end
     end
 
     -- Note that the logic inside the sync() function, i.e. privileges dump
```

It looks quite simple in terms of implementation, all we really need to do it to drop all "postponed_grant" issues on a sync and issue new warnings after sync iteration (per user).

Reply:
On the other hand, it scares me, how much work the program has to do on each small sync with possibly just a few grants/revokes.

Reply:
A brief summary of our offline discussion with Gleb: I'm going to implement this proposal. I also plan to post here a summary of cons and pros I see.

Reply:
I've pushed the new implementation.

Reply:
Thank you!